From 9a2ae8df1030046c94aa74bbd853dde5d844812d Mon Sep 17 00:00:00 2001
From: Hongxu Jia <hongxu.jia@windriver.com>
Date: Fri, 28 Feb 2025 19:13:18 -0800
Subject: [PATCH] cmd/go: clear GOROOT for func ldShared when -trimpath is used

In commit [cmd: remove support for GOROOT_FINAL][1], it clear GOROOT
for func ld when -trimpath is used. [2]

This commit do the same thing for func ldShared, otherwise run go_bootstrap
with -trimpath does not remove build path directory from the generated share
libarary

  $ ./make.bash --target-only --no-banner std
  $ go_bootstrap install -linkshared -buildmode=shared -trimpath std
  $ grep "$(pwd)" ./pkg/linux_amd64_dynlink/libstd.so
  Binary file ./pkg/linux_amd64_dynlink/libstd.so matches

[1] https://github.com/golang/go/commit/507d1b22f4b58ac68841582d0c2c0ab6b20e5a98
[2] https://github.com/golang/go/commit/507d1b22f4b58ac68841582d0c2c0ab6b20e5a98#diff-cab5921f94f2667bb0bc1b935d2d46b4c03541b4351b33438ab7290b94dea212R669

Upstream-Status: Submitted [https://go-review.googlesource.com/c/go/+/653895]

Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>
---
 src/cmd/go/internal/work/gc.go | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/src/cmd/go/internal/work/gc.go b/src/cmd/go/internal/work/gc.go
index 49b26e15609b..7c0d2ac9e0f0 100644
--- a/src/cmd/go/internal/work/gc.go
+++ b/src/cmd/go/internal/work/gc.go
@@ -721,6 +721,12 @@ func (gcToolchain) ldShared(b *Builder, root *Action, toplevelactions []*Action,
 	// the output file path is recorded in the .gnu.version_d section.
 	dir, targetPath := filepath.Split(targetPath)
 
+	// When -trimpath is used, GOROOT is cleared
+	if cfg.BuildTrimpath {
+		cfgChangedEnv = append(cfgChangedEnv, "GOROOT=")
+	} else {
+		cfgChangedEnv = append(cfgChangedEnv, "GOROOT="+cfg.GOROOT)
+	}
 	return b.Shell(root).run(dir, targetPath, cfgChangedEnv, cfg.BuildToolexec, base.Tool("link"), "-o", targetPath, "-importcfg", importcfg, ldflags)
 }
 
